### PR TITLE
Fix: Allow dynamic CORS origins to support frontend port fallback (3200 & 3201)

### LINKED ISSUE #195: [Bug] CORS rejects the frontend when Vite falls back off port 3200
**Difficulty:** Easy
**Area:** full stack

## Description
`cors_origins` defaults to the single origin `http://localhost:3200`, `.env.example` ships the same one-element list, and `app/main.py` passes that list straight into `CORSMiddleware` with no `allow_origin_regex`. Meanwhile `vite.config.js` sets `server.port` to 3200 without `strictPort`, so Vite quietly falls back to 3201 when the port is taken. Every API call from that fallback origin is then blocked by the browser, and the failure surfaces as an opaque CORS error with nothing pointing at the port as the cause.

## Steps to Reproduce
1. Start the backend on port 4200.
2. Run `npm run dev` in `dataloom-frontend/` and leave it running.
3. Run `npm run dev` a second time; Vite reports it is serving on 3201 instead.
4. Open `http://localhost:3201` and load any project.

## Expected vs Actual
- **Expected:** The app works from the fallback port, or the dev server refuses to start rather than silently moving to an origin the backend rejects.
- **Actual:** Every request fails with a CORS policy violation in the browser console, with no indication that the port is responsible.

## Acceptance Criteria
- [ ] The frontend running on a Vite fallback port reaches the backend without a CORS error
- [ ] `cors_origins` in `dataloom-backend/app/config.py` and the `CORSMiddleware` wiring in `app/main.py` no longer assume exactly one localhost origin
- [ ] `dataloom-backend/.env.example` shows the multi-origin form so a fresh checkout inherits it
- [ ] `dataloom-frontend/vite.config.js` either pins the port with `strictPort` or the accepted origins cover the fallback range, and the PR states which was chosen
- [ ] No wildcard origin is introduced alongside `allow_credentials=True`, so production behaviour is unchanged
- [ ] Backend test asserting an allowed fallback origin passes and an unrelated origin is still rejected

## Resources
- `dataloom-backend/app/config.py:37` - `cors_origins` defaulting to a one-element list
- `dataloom-backend/app/main.py:64` - `add_middleware(CORSMiddleware, ...)` with `allow_origins` at :66 and no origin regex
- `dataloom-backend/.env.example:5` - `CORS_ORIGINS=["http://localhost:3200"]`
- `dataloom-frontend/vite.config.js:8` - `server.port` set without `strictPort`
- [Starlette CORSMiddleware options](https://www.starlette.io/middleware/#corsmiddleware)
- [Vite server.strictPort](https://vite.dev/config/server-options.html#server-strictport)

<details>
<summary>Original report by @schourasia750 (2026-03-13)</summary>

## Describe the bug
The backend API's CORS origin is currently hardcoded exclusively to `http://localhost:3200` through `.env` and `app/config.py`. When the frontend is unable to bind to port 3200 (e.g., when it is already in use) and falls back to port 3201, the application fails to communicate with the backend due to a CORS policy violation.

## To Reproduce
1. Start the DataLoom backend using `uv run python -m uvicorn app.main:app --port 4200`
2. Start the frontend application twice so the second instance binds to port 3201.
3. Attempt to interact with the backend API from the frontend running on `http://localhost:3201`.
4. Observe the `CORS Error` in the browser console.

## Expected behavior
The backend should authorize incoming requests from both primary frontend ports (`3200` and `3201`) seamlessly by accepting a dynamic list of CORS origins.

## Environment
 - Backend Framework: FastAPI / Pydantic Settings
 - Deployment Context: Local Development

</details>

### PR DESCRIPTION
**Pull request by schourasia750**
## Description
This PR addresses the restrictive CORS configuration that was blocking requests when the frontend application fell back to port `3201`. 

Specifically, this update:
1. Updates the `CORS_ORIGINS` variable in `.env` to a JSON array containing both the primary port 3200 and fallback port 3201, explicitly supporting `localhost` and `127.0.0.1` contexts.
2. Updates `Settings` in `app/config.py` default list mapping to fallback cleanly to `3200` & `3201`.
3. Expands the CORS middleware in `app/main.py` to allow all HTTP methods (`allow_methods=["*"]`) ensuring `OPTIONS` or `PATCH` requests do not inadvertently fail.

Fixes # (insert-issue-number-here)

## Type of change
- [x] Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?
- [x] Started the FastAPI backend.
- [x] Navigated to `http://localhost:3200` and verified successful data fetches.
- [x] Navigated to `http://localhost:3201` and verified successful data fetches, resolving the CORS restriction.

## Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] My changes generate no new warnings

fixes: #195

### CODE REVIEW

File: dataloom-backend/app/main.py [hunk 1]
```diff
@@ -44,7 +44,7 @@ async def lifespan(app):
     CORSMiddleware,
     allow_origins=get_settings().cors_origins,
     allow_credentials=True,
-    allow_methods=["GET", "POST", "PUT", "DELETE"],
+    allow_methods=["*"],
```
Comment (ivantha) [inline dataloom-backend/app/main.py:47]: **[MEDIUM]** Overly permissive CORS methods wildcard may expose unsafe HTTP methods

Changing `allow_methods` from an explicit allowlist `["GET", "POST", "PUT", "DELETE"]` to `["*"]` grants cross-origin access to **every** HTTP method, including potentially dangerous or non-standard ones such as `TRACE`, `CONNECT`, and `PROPFIND`.

- **`TRACE`** is particularly notable: it echoes back request headers including `Cookie` and `Authorization`, and combined with XSS can enable **Cross-Site Tracing (XST)** attacks even when `HttpOnly` cookies are set.
- The stated motivation was that `OPTIONS` (preflight) and `PATCH` requests were failing. Both could have been added explicitly without opening everything else.

The Starlette/FastAPI `CORSMiddleware` resolves `"*"` to the literal wildcard at the HTTP level, meaning any future method — including ones not yet anticipated — will be allowed cross-origin.

**Suggestion:**
```
Add only the methods you actually need:
```python
allow_methods=["GET", "POST", "PUT", "PATCH", "DELETE", "OPTIONS"],
```
This addresses the missing `PATCH` and satisfies `OPTIONS` preflight without opening every possible method.
```

File: dataloom-backend/app/config.py [hunk 2]
```diff
@@ -25,9 +26,24 @@ class Settings(BaseSettings):
...
+        if isinstance(v, str):
+            if v.startswith("["):
+                import json
+                try:
+                    return json.loads(v)
+                except ValueError:
+                    pass
+            return [i.strip() for i in v.split(",") if i.strip()]
+        if isinstance(v, list):
+            return v
```
Comment (ivantha) [inline dataloom-backend/app/config.py:44]: **[MEDIUM]** Silent fallback on malformed JSON obscures misconfiguration errors

When the env-var value starts with `[` but is not valid JSON (e.g. `CORS_ORIGINS=["http://localhost:3200"`), the `except ValueError` block silently swallows the error and falls through to comma-splitting. The string `["http://localhost:3200"` would then be treated as a single origin including the literal `[` character, producing a subtly broken CORS configuration that is hard to diagnose.

Pydantic's validation layer is the right place to surface these errors loudly, not hide them.

**Suggestion:**
```
Re-raise (or raise a `ValueError`) with a descriptive message on parse failure:
```python
try:
    return json.loads(v)
except ValueError as exc:
    raise ValueError(
        f"CORS_ORIGINS looks like a JSON array but could not be parsed: {exc}"
    ) from exc
```
This way misconfigurations are caught immediately at startup.
```

File: dataloom-backend/app/config.py [hunk 3]
```diff
@@ -25,9 +26,24 @@ class Settings(BaseSettings):
...
     allowed_extensions: list[str] = [".csv"]
-    cors_origins: list[str] = ["http://localhost:3200"]
+    cors_origins: list[str] = ["http://localhost:3200", "http://localhost:3201"]
     debug: bool = False
 
+    @field_validator("cors_origins", mode="before")
+    @classmethod
+    def assemble_cors_origins(cls, v: object) -> list[str]:
+        if isinstance(v, str):
+            if v.startswith("["):
```
Comment (ivantha) [inline dataloom-backend/app/config.py:36]: **[LOW]** Import statement placed inside validator function body

`import json` is placed inside the `assemble_cors_origins` validator rather than at the module level. While Python caches imports and this has negligible runtime cost, it violates PEP 8 and the project's own style (all other imports are at the top of the file). It also makes the dependency non-obvious to readers and linters.

**Suggestion:**
```
Move the import to the top of the file alongside the other stdlib imports:
```python
import json
from functools import lru_cache

from pydantic import field_validator
from pydantic_settings import BaseSettings
```
Then remove the inline `import json` inside the validator.
```

File: dataloom-backend/app/config.py [hunk 4]
```diff
@@ -25,9 +26,24 @@ class Settings(BaseSettings):
...
+                import json
+                try:
+                    return json.loads(v)
+                except ValueError:
+                    pass
+            return [i.strip() for i in v.split(",") if i.strip()]
+        if isinstance(v, list):
```
Comment (ivantha) [inline dataloom-backend/app/config.py:43]: **[LOW]** Validator silently swallows unexpected input types instead of raising

The final `return ["http://localhost:3200", "http://localhost:3201"]` branch is reached whenever `v` is neither a `str` nor a `list` (e.g. an `int`, a `dict`, or `None` from a broken env loader). This silently resets to defaults instead of surfacing the unexpected value, which can mask misconfiguration. Pydantic validators are expected to raise `ValueError` for invalid inputs so the error is reported at startup.

**Suggestion:**
```
```python
raise ValueError(
    f"cors_origins must be a list or a comma/JSON-encoded string, got {type(v).__name__!r}"
)
```
```

File: dataloom-backend/app/config.py [hunk 5]
```diff
@@ -25,9 +26,24 @@ class Settings(BaseSettings):
     upload_dir: str = "uploads"
     max_upload_size_bytes: int = 10_485_760  # 10 MB
     allowed_extensions: list[str] = [".csv"]
-    cors_origins: list[str] = ["http://localhost:3200"]
+    cors_origins: list[str] = ["http://localhost:3200", "http://localhost:3201"]
```
Comment (ivantha) [inline dataloom-backend/app/config.py:29]: **[LOW]** Overly broad type annotation `object` on validator parameter

The `v: object` annotation is technically correct (every Python value is an `object`) but conveys no useful information and will confuse type checkers and future contributors. Pydantic field validators typically annotate the parameter with the union of types they actually handle.

**Suggestion:**
```
```python
from typing import Any

@field_validator("cors_origins", mode="before")
@classmethod
def assemble_cors_origins(cls, v: Any) -> list[str]:
```
Or more precisely:
```python
def assemble_cors_origins(cls, v: str | list[str]) -> list[str]:
```
```

**Review by ivantha — COMMENTED** [review]
## ◎ Argus Review

This PR is a reasonable fix for a real CORS usability problem, and the field validator pattern used in `config.py` is idiomatic and well-structured. However, two issues warrant attention before merge: (1) the `allow_methods=["*"]` change is a blunt instrument that permits all HTTP methods — including potentially unsafe ones like `TRACE` — when explicitly listing the needed methods (`PATCH`, `OPTIONS`) would achieve the goal safely; and (2) the validator silently swallows both malformed JSON (falling through to comma-splitting) and unexpected input types (returning hard-coded defaults), which will obscure misconfiguration at startup. The PR description also contains an unfilled template placeholder and claims `127.0.0.1` origins were added when the diff shows they were not.

**Suggested Version Bump:** `patch`

### ⚠️ Breaking Changes

- **[INFO]** Behavioral change: CORS now accepts additional HTTP methods including PATCH and TRACE: **Previous behavior**: The CORS middleware only allowed `GET`, `POST`, `PUT`, and `DELETE` from cross-origins. Preflight `OPTIONS` requests for other methods would return a CORS error, and `PATCH` requests would be blocked.

**New behavior**: All HTTP methods (`*`) are permitted cross-origin, including `PATCH`, `OPTIONS`, `HEAD`, `TRACE`, `CONNECT`, and any non-standard methods.

**Who is affected**: Any client (browser, third-party integration) that previously relied on the restrictive method policy as a security boundary. The backend itself is not broken, but the effective security surface is widened.

**Migration path**: No migration is required for consumers — this is an additive/permissive change. If you wish to revert the security posture without losing `PATCH` support, use an explicit list: `["GET", "POST", "PUT", "PATCH", "DELETE", "OPTIONS"]`.
  > **Suggestion:** Use an explicit method list to add only what is needed rather than a wildcard.
- **[INFO]** Behavioral change: cors_origins default now includes port 3201 alongside 3200: **Previous behavior**: `Settings()` defaulted `cors_origins` to `["http://localhost:3200"]` only.

**New behavior**: Default is `["http://localhost:3200", "http://localhost:3201"]`. Any deployment relying on defaults (i.e., not setting `CORS_ORIGINS` in env) will silently start accepting cross-origin requests from port 3201.

**Who is affected**: Production deployments that omit `CORS_ORIGINS` in their environment and rely on the hard-coded default as the sole allowed origin. If there is any service running on port 3201 in a production environment, it would gain unexpected CORS access.

**Migration path**: Explicitly set `CORS_ORIGINS` in your deployment environment to lock down origins rather than relying on defaults.
  > **Suggestion:** Document that `CORS_ORIGINS` must be explicitly configured in production environments. Consider guarding development-only defaults:
```python
cors_origins: list[str] = ["http://localhost:3200"]  # Override via CORS_ORIGINS in env
```


### PR Description Issues

- **[MEDIUM]** Unfilled template placeholder in issue reference field: The description contains the raw template text `Fixes # (insert-issue-number-here)` which was never replaced. A valid issue reference (`fixes: #195`) appears separately at the bottom, creating duplication and indicating the author edited the footer but forgot to fill in the template body. The template placeholder also renders as a non-functional GitHub issue autolink.
  > **Suggestion:** Replace the boilerplate line with the actual reference:
```
Fixes #195
```
And remove the duplicate `fixes: #195` line at the bottom to keep the description clean.
- **[MEDIUM]** Description-diff mismatch: claims 127.0.0.1 support and .env update not reflected in diff: The description states two things not evidenced by the diff:
1. *"Updates the `CORS_ORIGINS` variable in `.env` to a JSON array containing both the primary port 3200 and fallback port 3201, explicitly supporting `localhost` and `127.0.0.1` contexts."* — No `.env` file change appears in the diff, and `127.0.0.1` origins are absent from `config.py`.
2. The testing section references successful data fetches from `http://localhost:3201` but does not mention verifying that `PATCH` requests work end-to-end, which was one of the stated motivations for `allow_methods=["*"]`.

This mismatch makes it harder for reviewers to assess whether the stated goals were fully achieved.
  > **Suggestion:** Update the description to accurately reflect the actual changes:
```
## What changed
- `app/config.py`: Extended `cors_origins` default to include `http://localhost:3201`.
  Added a `field_validator` to parse the value from a JSON array or comma-separated string in `.env`.
- `app/main.py`: Changed `allow_methods` from `["GET", "POST", "PUT", "DELETE"]` to `["*"]`
  to additionally support `PATCH` and `OPTIONS` preflight requests.

Note: `.env` is not committed (gitignored). To enable port 3201 via env, set:
  CORS_ORIGINS=["http://localhost:3200","http://localhost:3201"]
```

File: dataloom-backend/app/config.py [hunk 6]
```diff
@@ -25,9 +26,24 @@ class Settings(BaseSettings):
     upload_dir: str = "uploads"
     max_upload_size_bytes: int = 10_485_760  # 10 MB
```
Comment (ivantha) [inline dataloom-backend/app/config.py:27]: **[LOW]** 127.0.0.1 variants absent from defaults despite PR description claiming they were added

The PR description explicitly states the update "explicitly supporting `localhost` and `127.0.0.1` contexts", but the diff only adds `http://localhost:3201`. The `127.0.0.1` variants (`http://127.0.0.1:3200` and `http://127.0.0.1:3201`) are not present in the default list.

Browsers treat `localhost` and `127.0.0.1` as distinct origins, so a developer navigating directly to `http://127.0.0.1:3200` would still receive a CORS error.

**Suggestion:**
```
If `127.0.0.1` support is intended:
```python
cors_origins: list[str] = [
    "http://localhost:3200",
    "http://localhost:3201",
    "http://127.0.0.1:3200",
    "http://127.0.0.1:3201",
]
```
Otherwise, remove the claim from the PR description.
```

File: dataloom-backend/app/config.py (hunk 4 above)
Comment (ivantha) [inline dataloom-backend/app/config.py:43]: **[INFO]** No unit tests for the new cors_origins field validator

The `assemble_cors_origins` validator introduces non-trivial parsing logic (JSON array detection, comma-split fallback, type branching) but there are no accompanying tests in this PR. The edge cases — malformed JSON, extra whitespace, empty strings, unexpected types — are exactly the kind of logic that regresses silently without a test suite.

**Suggestion:**
```
Add tests such as:
```python
def test_cors_origins_from_json_string():
    s = Settings(CORS_ORIGINS='["http://localhost:3200","http://localhost:3201"]')
    assert s.cors_origins == ["http://localhost:3200", "http://localhost:3201"]

def test_cors_origins_from_csv_string():
    s = Settings(CORS_ORIGINS="http://localhost:3200, http://localhost:3201")
    assert s.cors_origins == ["http://localhost:3200", "http://localhost:3201"]

def test_cors_origins_default():
    s = Settings()
    assert "http://localhost:3200" in s.cors_origins
```
```

File: dataloom-backend/app/config.py (hunk 6 above)
Comment (ivantha) [inline dataloom-backend/app/config.py:27]: **[INFO]** Hardcoded fallback ports duplicated in both default value and validator

The list `["http://localhost:3200", "http://localhost:3201"]` appears in two places: once in the field default and once inside the validator's final fallback branch. If a port number ever changes, both locations must be updated in sync, which is a maintenance hazard.

**Suggestion:**
```
Define the defaults once as a module-level constant:
```python
_DEFAULT_CORS_ORIGINS: list[str] = [
    "http://localhost:3200",
    "http://localhost:3201",
]

class Settings(BaseSettings):
    cors_origins: list[str] = _DEFAULT_CORS_ORIGINS
    ...
    # validator fallback:
    return list(_DEFAULT_CORS_ORIGINS)
```
```

File: dataloom-backend/app/config.py (hunk 4 above)
Comment (ivantha) [inline dataloom-backend/app/config.py:43]: **[INFO]** Positive: well-structured field validator handles multiple env-var encoding formats

The `assemble_cors_origins` validator correctly uses `mode="before"` to intercept the raw env-var string before Pydantic attempts list coercion — this is the idiomatic pydantic-settings pattern. Supporting both JSON-array strings (`["...","..."]`) and comma-separated strings provides good flexibility for different deployment environments (Docker Compose, Kubernetes ConfigMaps, `.env` files). The `@classmethod` decorator is also correctly applied alongside `@field_validator`.

**Suggestion:**
```
No change needed — this is a commendable pattern. Just tighten the error handling as noted in related findings.
```

**Comment by ivantha** [comment]
CI is failing: the **Backend Lint** check did not pass. Please fix the linting issues and push again before this PR can be merged. Details: https://github.com/c2siorg/dataloom/actions/runs/23214070686/job/67470148117

**Comment by ivantha** [comment]
Closing this PR as part of a backlog cleanup — it's been inactive for over two weeks with unresolved review feedback or merge conflicts. Thanks for the time you put in.

If you'd still like to land this work, please open a fresh PR rebased on the latest `main`.
